File: python/src/tools/dashboards/update.py
import json
from pydantic import BaseModel
from src.tools.types import Context, Tool, ToolResult, TextContent
from src.schema.dashboards import UpdateDashboardInput
from src.lib.utils.api import get_project_base_url


class UpdateDashboardParams(BaseModel):
    dashboardId: int
    data: UpdateDashboardInput


async def update_dashboard_handler(context: Context, params: UpdateDashboardParams) -> ToolResult:
    project_id = await context.get_project_id()
    dashboard_result = await context.api.dashboards(project_id).update(params.dashboardId, params.data)

    if not dashboard_result.success:
        raise Exception(f"Failed to update dashboard: {dashboard_result.error}")

    dashboard_with_url = {
        **dashboard_result.data.model_dump(),
        "url": f"{get_project_base_url(project_id)}/dashboard/{dashboard_result.data.id}"
    }

    return ToolResult(content=[TextContent(text=json.dumps(dashboard_with_url))])


def update_dashboard_tool() -> Tool[UpdateDashboardParams]:
    return Tool(
        name="dashboard-update",
        description="""
        - Update an existing dashboard by ID.
        - Can update name, description, pinned status or tags.
        """,
        schema=UpdateDashboardParams,
        handler=update_dashboard_handler
    )

File: typescript/src/tools/dashboards/update.ts
import { DashboardUpdateSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = DashboardUpdateSchema;

type Params = z.infer<typeof schema>;

export const updateHandler = async (context: Context, params: Params) => {
	const { dashboardId, data } = params;
	const projectId = await context.stateManager.getProjectId();
	const dashboardResult = await context.api
		.dashboards({ projectId })
		.update({ dashboardId, data });

	if (!dashboardResult.success) {
		throw new Error(`Failed to update dashboard: ${dashboardResult.error.message}`);
	}

	const dashboardWithUrl = {
		...dashboardResult.data,
		url: `${context.api.getProjectBaseUrl(projectId)}/dashboard/${dashboardResult.data.id}`,
	};

	return { content: [{ type: "text", text: JSON.stringify(dashboardWithUrl) }] };
};

const tool = (): Tool<typeof schema> => ({
	name: "dashboard-update",
	title: getToolDefinition("dashboard-update").title,
	description: getToolDefinition("dashboard-update").description,
	schema,
	handler: updateHandler,
	annotations: {
		destructiveHint: false,
		idempotentHint: true,
		openWorldHint: true,
		readOnlyHint: false,
	},
});

export default tool;
